src/transformStateWithClones.js

```diff
@@ -7,7 +7,33 @@
  * @return {Object[]}
  */
 function transformStateWithClones(state, actions) {
-  // write code here
+  let stateCopy = { ...state };
+  const states = [];
+
+  for (const action of actions) {
+    switch (action.type) {
+      case 'clear':
+        stateCopy = {};
+        break;
+
+      case 'addProperties':
+        stateCopy = { ...stateCopy, ...action.extraData };
+        break;
+
+      case 'removeProperties':
+        for (const key of action.keysToRemove) {
+          delete stateCopy[key];
+        }
+        break;
+
+      default:
+        throw new Error(`Unknown action type: ${action.type}`);
+    }
+
+    states.push({ ...stateCopy });
```

Comment:
[DONT REPEAT YOURSELF]: You are creating a copy of the `stateCopy` object every time you push it to the `states` array. Since you are already creating a new `stateCopy` at the start of each action (lines 16, 20), there is no need to create an additional copy here. You can simply push the `stateCopy` directly to the `states` array.